### ISSUE TITLE
What about breaking changes?

### ISSUE THREAD

**Issue by bryant1410** [mahmoud/calver#4]
I think this versioning scheme should contemplate breaking changes.

A very important feature I see in today's common approaches such as semver is that they reflect the breaking changes in their numbering scheme. However, I see there is nothing similar stated in this specification. Shouldn't it be contemplated?

I would be concerned if I have several dependencies using this scheme and I want to update them. Will certain updates break my code? I will have to check and remember each documentation to find which ones break compatibility (for instance, twisted states that the first two numbers indicate the API compatibility).

**Comment by mahmoud** [comment]
It's not at all just JavaScript, I can't even begin to enumerate the times I've seen this happen.

Depending on the scale of the API break, the right thing to do is rename the project. From Zope3/BlueBream to characteristic/attrs, this strategy has worked out best for new users and old. It may not be good for building the #brand, but it's best for the ecosystem.

If someone wants to use a character or two after the version number to indicate that a major breaking change occurred, it can still be CalVer, but I would find it a stretch to make that a recommendation. Besides, CalVer is a mostly descriptive "specification", so there would have to be some actual examples to merit putting the practice on the site.

**Comment by bryant1410** [comment]
> It's not at all just JavaScript, I can't even begin to enumerate the times I've seen this happen.
> 
> Depending on the scale of the API break, the right thing to do is rename the project. From Zope3/BlueBream to characteristic/attrs, this strategy has worked out best for new users and old. It may not be good for building the #brand, but it's best for the ecosystem.

Yeah, in Java too: [Java Interoperability Policy for Major Version Updates](https://jakewharton.com/java-interoperability-policy-for-major-version-updates/)

> 
> If someone wants to use a character or two after the version number to indicate that a major breaking change occurred, it can still be CalVer, but I would find it a stretch to make that a recommendation. Besides, CalVer is a mostly descriptive "specification", so there would have to be some actual examples to merit putting the practice on the site.

**Comment by mahmoud** [comment]
Just to expand on that last comment a little more: I've seen this happen in JavaScript and Python, but I think frequency is correlated with the high-level-ness of the interface. That is, I've seen this happen most in interfaces, both APIs and UIs.

Because there's no consensus with what constitutes a breaking change in an API (e.g., is adding a field a breaking change? is changing the error message on error responses a breaking change?), the only way to get some form of breaking change semantics would be to have extensive documentation. This defeats the purpose of SemVer (communication-less announcement of breaking APIs), and strengthens that of CalVer (encouraging documentation and communication).

Sounds like I may need to create a FAQ for the calver.org site :)

**Comment by peterjc** [comment]
I mentioned another hybrid example on #5 using what I would call ``MAJOR.0Y.0M.0D`` in the CalVer style notation.
